p2pclient/daemon.py

```diff
@@ -0,0 +1,234 @@
+
+    @abc.abstractmethod
+    def _make_command_line_options(self) -> List[str]:
+        ...
+
+    @abc.abstractmethod
+    def _terminate(self) -> None:
+        ...
+
+    def _run(self, daemon_executable: str):
+        cmd_list = [daemon_executable] + self._make_command_line_options()
+        self.proc_daemon = subprocess.Popen(
+            cmd_list, stdout=self.f_log, stderr=self.f_log, bufsize=0
+        )
+
+    async def wait_until_ready(self):
+        lines_head_occurred = {line: False for line in self.LINES_HEAD_PATTERN}
+
+        with open(self.log_filename, "rb") as f_log_read:
+
+            async def read_from_daemon_and_check():
+                line = f_log_read.readline()
+                for head_pattern in lines_head_occurred:
+                    if line.startswith(head_pattern):
+                        lines_head_occurred[head_pattern] = True
+                return all([value for _, value in lines_head_occurred.items()])
+
+            await try_until_success(read_from_daemon_and_check)
+
+        # sleep for a while in case that the daemon haven't been ready after emitting these lines
+        await anyio.sleep(0.1)
+
+    def close(self):
+        if self.is_closed:
+            return
+        self._terminate()
+        self.proc_daemon.wait()
+        self.f_log.close()
+        self.is_closed = True
+
+
+class GoDaemon(Daemon):
+
+    LINES_HEAD_PATTERN = (b"Control socket:", b"Peer ID:", b"Peer Addrs:")
+
+    def _make_command_line_options(self) -> List[str]:
+        cmd_list = [f"-listen={str(self.control_maddr)}"]
+        if self.enable_connmgr:
+            cmd_list += ["-connManager=true", "-connLo=1", "-connHi=2", "-connGrace=0"]
+        if self.enable_dht:
+            cmd_list += ["-dht=true"]
+        if self.enable_pubsub:
+            cmd_list += ["-pubsub=true", "-pubsubRouter=gossipsub"]
+
+        return cmd_list
+
+    def _terminate(self) -> None:
+        self.proc_daemon.terminate()
+
+
+class JsDaemon(Daemon):
+
+    LINES_HEAD_PATTERN = (b"daemon has started",)
+
+    def _make_command_line_options(self) -> List[str]:
+        cmd_list = [f"--listen={str(self.control_maddr)}"]
+        if self.enable_connmgr:
+            cmd_list += [
+                "--connManager=true",
+                "--connLo=1",
+                "--connHi=2",
+                "--connGrace=0",
+            ]
+        if self.enable_dht:
+            cmd_list += ["--dht=true"]
+        if self.enable_pubsub:
+            cmd_list += ["--pubsub=true", "--pubsubRouter=gossipsub"]
+
+        return cmd_list
+
+    # TODO: investigate why the JS daemon needs to be killed instead of terminating gracefully. Some tests
+    #       (ex: test_client_stream_open_failure) freeze after completion if we use terminate.
+    def _terminate(self) -> None:
+        self.proc_daemon.kill()
+
+
+class DaemonTuple(NamedTuple):
+    daemon: Daemon
+    client: Client
+
+
+@asynccontextmanager
+async def make_p2pd_pair_unix(
```

Comment:
IMO `make_p2pd_pair_unix` and `make_p2pd_pair_ip4` should stay in `tests` since they have maddrs hardcoded for testing, while `_make_p2pd_pair` should stay in `p2pclient/daemon.py` and probably renamed to `make_p2pd_pair()`, as a utililty function for user to spawn a daemon and control it. Does it make sense?

```python
async with make_p2pd_pair() as p2pdc:
    client = p2pdc.client
    ...
```

Reply:
`make_p2pd_pair` can probably be optimized for UX by allowing parameters to be optional and given default values. But that can be discussed later.

Reply:
That makes sense, however the main usage of the daemon is for testing. Maybe we can provide them as utility functions, and make make_p2pd_pair a public function of the interface as well?

Reply:
Sounds good:)